src/librustc/middle/traits/error_reporting.rs

```diff
@@ -62,6 +66,73 @@ pub fn report_projection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
     }
 }
 
+fn report_on_unimplemented<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
+                                     trait_ref: &TraitRef<'tcx>) -> Option<String> {
+    let def_id = trait_ref.def_id;
+    let mut report = None;
+    ty::each_attr(infcx.tcx, def_id, |item| {
+        if item.check_name("on_unimplemented") {
+            match item.meta().node {
```

Comment:
This pair of nested `match`s can be `if let Some(ref i) = item.value_str() { ... }`.